### PR TITLE
ccl,sql,util: Fix jwt auth and add sensitive error logs

### LINKED ISSUE #123575: jwtauthccl : http client missing support for proxies
**Describe the problem**

The new JWKS Auto fetch feature introduced through https://github.com/cockroachdb/cockroach/pull/117054 ignored http transport env vars like HTTP_PROXY. As a result, the outbound connection to fetch jwks url would fail. 

**To Reproduce**

Set `server.jwt_authentication.jwks_auto_fetch.enabled` cluster setting to True in an environment with HTTP_PROXY setup. 

**Expected behavior**
The JWT Auth should work as expected, but it returns error: "JWT authentication: unable to validate token"

**Environment:**
 - CockroachDB version : v23.1.17, v24.1 

**Additional context**
This is similar to https://github.com/cockroachdb/cockroach/issues/32803

Jira issue: CRDB-38408

gz#21355

Related to : https://github.com/cockroachdb/cockroach/issues/123605

### PR DESCRIPTION
**Pull request by souravcrl**
ccl,sql,util: Fix jwt auth and add sensitive error logs

We are running into issues with jwt authentication and currently unable to
provide support as we are not logging the error from the http client used in the
authenticator. The PR looks to propagate this obtained error from
`ValidateJWTLogin` http client. The http client now also respects the system
http proxy if set.

Validated the error details when presenting an expired token
```
ERROR: JWT authentication: invalid token
SQLSTATE: 28000
DETAIL: unable to parse token: exp not satisfied Failed running "sql"
```

Validated error on setting wrong proxy params
```
ERROR: JWT authentication: unable to validate token
SQLSTATE: 28000
Failed running "sql"
```
and logged error:
```
I240510 08:31:28.604141 1473 4@util/log/event_log.go:32 ⋮
[T1,Vsystem,n1,client=127.0.0.1:56289,hostssl,user=‹sourav.sarangi›] 3
={"Timestamp":1715329888604122000,"EventType":"client_authentication_failed","InstanceID":1,"Network":"tcp","RemoteAddress":"‹127.0.0.1:56289›","SessionID":"17ce136f2a8ecd480000000000000001","Transport":"hostssl","User":"‹sourav.sarangi›","SystemIdentity":"‹sourav.sarangi›","Reason":"CREDENTIALS_INVALID","Detail":"JWT
authentication: unable to validate token\nunable to fetch jwks: Get
\"https://accounts.google.com/.well-known/openid-configuration\": proxyconnect
tcp: dial tcp [::1]:3129: connect: connection refused","Method":"jwt_token"}
```

Verified access logs after setting up squid proxy and passing env HTTP_PROXY and
HTTPS_PROXY params
```
1715103871.761    144 ::1 TCP_TUNNEL/200 5708 CONNECT accounts.google.com:443 -
HIER_DIRECT/74.125.200.84 -
1715103871.836     73 ::1 TCP_TUNNEL/200 5964 CONNECT www.googleapis.com:443 -
HIER_DIRECT/142.250.182.10 -
```

fixes https://github.com/cockroachdb/cockroach/issues/123575, CRDB-38386, CRDB-38408
Epic None

Release note: Noneccl,sql,util: Fix jwt auth and add sensitive error logs

We are running into issues with jwt authentication and currently unable to
provide support as we are not logging the error from the http client used in the
authenticator. The PR looks to propagate this obtained error from
`ValidateJWTLogin` http client. The http client now also respects the system
http proxy if set.

Validated the error details when presenting an expired token
```
ERROR: JWT authentication: invalid token
SQLSTATE: 28000
DETAIL: unable to parse token: exp not satisfied Failed running "sql"
```

Validated error on setting wrong proxy params
```
ERROR: JWT authentication: unable to validate token
SQLSTATE: 28000
Failed running "sql"
```
and logged error:
```
I240510 08:31:28.604141 1473 4@util/log/event_log.go:32 ⋮
[T1,Vsystem,n1,client=127.0.0.1:56289,hostssl,user=‹sourav.sarangi›] 3
={"Timestamp":1715329888604122000,"EventType":"client_authentication_failed","InstanceID":1,"Network":"tcp","RemoteAddress":"‹127.0.0.1:56289›","SessionID":"17ce136f2a8ecd480000000000000001","Transport":"hostssl","User":"‹sourav.sarangi›","SystemIdentity":"‹sourav.sarangi›","Reason":"CREDENTIALS_INVALID","Detail":"JWT
authentication: unable to validate token\nunable to fetch jwks: Get
\"https://accounts.google.com/.well-known/openid-configuration\": proxyconnect
tcp: dial tcp [::1]:3129: connect: connection refused","Method":"jwt_token"}
```

Verified access logs after setting up squid proxy and passing env HTTP_PROXY and
HTTPS_PROXY params
```
1715103871.761    144 ::1 TCP_TUNNEL/200 5708 CONNECT accounts.google.com:443 -
HIER_DIRECT/74.125.200.84 -
1715103871.836     73 ::1 TCP_TUNNEL/200 5964 CONNECT www.googleapis.com:443 -
HIER_DIRECT/142.250.182.10 -
```

fixes https://github.com/cockroachdb/cockroach/issues/123575, CRDB-38386, CRDB-38408
Epic None

Release note: None

### CODE REVIEW

**Review by souravcrl — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/123697)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @dhartunian)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/123697#-NxJhxV85fcnbrJqS9O3:-NxJhxV85fcnbrJqS9O4:b-ydp83z) (no related file):*
Should we support cluster settings to control http client behaviour for jwt?  
Do we log err desc. (which contains sensitive info)  in db logs but redact in sentry? Also do we send err desc to client?

___

File: pkg/ccl/jwtauthccl/authentication_jwt.go [hunk 1]
```diff
@@ -179,7 +186,7 @@ func (authenticator *jwtAuthenticator) ValidateJWTLogin(
 	// Now that both the issuer and key-id are matched, parse the token again to validate the signature.
 	parsedToken, err := jwt.Parse(tokenBytes, jwt.WithKeySet(jwkSet), jwt.WithValidate(true), jwt.InferAlgorithmFromKey(true))
 	if err != nil {
-		return errors.Newf("JWT authentication: invalid token")
+		return errors.WithDetailf(errors.Newf("JWT authentication: invalid token"), "unable to parse token: %v", err)
```
Comment (BabuSrithar) [inline pkg/ccl/jwtauthccl/authentication_jwt.go:189]: I thought WithDetailf is going to send the error details all the way to client. With this change, looks like we are sending more details visible to client ?

**Review by dhartunian — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/123697)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @BabuSrithar, @bdarnell, and @souravcrl)

___
*[`pkg/ccl/jwtauthccl/authentication_jwt.go` line 189 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/123697#gh-1593988975:-NxNNfeUAeqWcZe4bpef:b-ngnpsi) ([raw file](https://github.com/cockroachdb/cockroach/blob/806b46e3d118c405261b78d2e459bb19356cc0ec/pkg/ccl/jwtauthccl/authentication_jwt.go#L189)):*
<details><summary><i>Previously, BabuSrithar (BabuSrithar) wrote…</i></summary><blockquote>

I thought WithDetailf is going to send the error details all the way to client. With this change, looks like we are sending more details visible to client ?
</blockquote></details>

What we did in the OIDC code was put the detailed error behind a conditional log using `log.V(1)` and keep the simple error for returning to client.

https://github.com/cockroachdb/cockroach/blob/44d7da514c7c0d21411fa31caaa59457b413540a/pkg/ccl/oidcccl/authentication_oidc.go#L343-L345

This way a customer can turn on the logs using vmodule when necessary for debugging and then turn them off again.

vmodule levels can be enabled via a flag to cockroach like: ` --vmodule=authentication_oidc=1` (It's based on filename) or via SQL on a running cluster like this: `SELECT crdb_internal.set_vmodule('authentication_oidc=1');`
___
*[`pkg/ccl/jwtauthccl/authentication_jwt_test.go` line 797 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/123697#-NxNOpv2Bpu1O_1hE_Fk:-NxNOpv2Bpu1O_1hE_Fl:b-pclp5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/806b46e3d118c405261b78d2e459bb19356cc0ec/pkg/ccl/jwtauthccl/authentication_jwt_test.go#L797)):*
> ```Go
> 	// happen to create an http client, then the DefaultTransport will have
> 	// been been initialized with an empty Proxy.  So, set proxy directly.
> 	http.DefaultTransport.(*http.Transport).Proxy = func(_ *http.Request) (*url.URL, error) {
> ```

try using `testutils.TestingHook` for this pattern just to keep things consistent.

**Review by bdarnell — COMMENTED** [review]
> Reviewed 1 of 8 files at r1, 4 of 7 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/123697)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @BabuSrithar and @souravcrl)

___
*[`-- commits` line 45 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/123697#-NxNT52A9FTol0K_fPyP):*
In general, we want to unify all the HTTP client usage in the system and not add new knobs that are specific to one use or another. We could add a custom_ca setting if we need it (following the example of cloudstorage.http.custom_ca) but we shouldn't have a new system_proxy setting that nothing else uses (the NO_PROXY env var can be use to control when the proxy is and is not used). 
___
*[`pkg/ccl/jwtauthccl/authentication_jwt.go` line 189 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/123697#gh-1593988975:-NxNUBPb3J-Y7hd1KKGk:bedp98) ([raw file](https://github.com/cockroachdb/cockroach/blob/806b46e3d118c405261b78d2e459bb19356cc0ec/pkg/ccl/jwtauthccl/authentication_jwt.go#L189)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

What we did in the OIDC code was put the detailed error behind a conditional log using `log.V(1)` and keep the simple error for returning to client.

https://github.com/cockroachdb/cockroach/blob/44d7da514c7c0d21411fa31caaa59457b413540a/pkg/ccl/oidcccl/authentication_oidc.go#L343-L345

This way a customer can turn on the logs using vmodule when necessary for debugging and then turn them off again.

vmodule levels can be enabled via a flag to cockroach like: ` --vmodule=authentication_oidc=1` (It's based on filename) or via SQL on a running cluster like this: `SELECT crdb_internal.set_vmodule('authentication_oidc=1');`
</blockquote></details>

Yes, we want to return limited data to the client, and specifically the client should not be able to learn any details about URLs or proxy setup from our error messages. I think errors relating to the network fetch should use log.Errorf (without log.V(1) - that just makes it harder to see because no one knows which vmodule options to turn on).

Parsing could go either way - in theory the parse error would not contain any information that the client didn't just give to us, so it would be safe to return as an error detail, but it's probably better to play it safe and leave the details in the log. This would be a better use for log.V(1) so that a malicious client can't easily spam the logs. 
___
*[`pkg/sql/pgwire/auth.go` line 556 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/123697#-NxNV1hBEfFduTXn7rPX:-NxNV1hBEfFduTXn7rPY:bez00vs) ([raw file](https://github.com/cockroachdb/cockroach/blob/806b46e3d118c405261b78d2e459bb19356cc0ec/pkg/sql/pgwire/auth.go#L556)):*
> ```Go
> 		var errStr redact.RedactableString
> 		if detailedErr != nil {
> 			// should we apply redact here?
> ```

I don't think so - redaction is about what can be sent to CRL. For error messages we generally want to either send the whole thing to the client or none of it. 
___
*[`pkg/sql/pgwire/auth_methods.go` line 789 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/123697#-NxNVQs850pBOLCDo3-8:-NxNVQs850pBOLCDo3-9:by09zbg) ([raw file](https://github.com/cockroachdb/cockroach/blob/806b46e3d118c405261b78d2e459bb19356cc0ec/pkg/sql/pgwire/auth_methods.go#L789)):*
> ```Go
> 			c.LogAuthFailed(ctx, eventpb.AuthFailReason_CREDENTIALS_INVALID, err)
> 			// we are returning unsafe error to client in description but only safe
> 			// log parts are sent to sentry. is this correct way to handle this?
> ```

Yes, it can be (although it depends on what the "unsafe" parts are). For example, usernames are generally marked as redacted and should not be sent to sentry, but they could be appropriate to return in an error message. 
___
*[`pkg/util/httputil/client.go` line 46 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/123697#-NxNVcazCCi3xE2K0tmE:-NxNVcazCCi3xE2K0tmF:b8bbswm) ([raw file](https://github.com/cockroachdb/cockroach/blob/806b46e3d118c405261b78d2e459bb19356cc0ec/pkg/util/httputil/client.go#L46)):*
> ```Go
> 	var tlsConf *tls.Config
> 	if customCAPem != "" {
> 		roots, err := x509.SystemCertPool()
> ```

I think custom CAs are generally used with a new empty cert pool, not SystemCertPool.

**Review by souravcrl — COMMENTED** [review]
PTAL. I have addressed the previous comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/123697#-:-NxTZa5R51iGiWHwwOmU:b54xn9k)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @BabuSrithar, @bdarnell, and @dhartunian)

___
*[`-- commits` line 45 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/123697#-NxNT52A9FTol0K_fPyP):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

In general, we want to unify all the HTTP client usage in the system and not add new knobs that are specific to one use or another. We could add a custom_ca setting if we need it (following the example of cloudstorage.http.custom_ca) but we shouldn't have a new system_proxy setting that nothing else uses (the NO_PROXY env var can be use to control when the proxy is and is not used).
</blockquote></details>

removed cluster settings `client.timeout` and `client.system_proxy.enabled`
___
*[`pkg/ccl/jwtauthccl/authentication_jwt.go` line 74 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/123697#gh-1593978758:-NxOLjsW4wQZqrifkcWY:bik99gc) ([raw file](https://github.com/cockroachdb/cockroach/blob/806b46e3d118c405261b78d2e459bb19356cc0ec/pkg/ccl/jwtauthccl/authentication_jwt.go#L74)):*
<details><summary><i>Previously, BabuSrithar (BabuSrithar) wrote…</i></summary><blockquote>

Commented code may be intended to be removed ? 
</blockquote></details>

removed commented code post discussion
___
*[`pkg/ccl/jwtauthccl/authentication_jwt.go` line 96 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/123697#gh-1593978866:-NxOLoTmCI2LDZ4GKAc4:bik99gc) ([raw file](https://github.com/cockroachdb/cockroach/blob/806b46e3d118c405261b78d2e459bb19356cc0ec/pkg/ccl/jwtauthccl/authentication_jwt.go#L96)):*
<details><summary><i>Previously, BabuSrithar (BabuSrithar) wrote…</i></summary><blockquote>

Commented code may be intended to be removed ? 
</blockquote></details>

removed commented code post discussion
___
*[`pkg/ccl/jwtauthccl/authentication_jwt.go` line 189 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/123697#gh-1593988975:-NxSZQFf2MaQfUcMhRZk:b-86fc3l) ([raw file](https://github.com/cockroachdb/cockroach/blob/806b46e3d118c405261b78d2e459bb19356cc0ec/pkg/ccl/jwtauthccl/authentication_jwt.go#L189)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yes, we want to return limited data to the client, and specifically the client should not be able to learn any details about URLs or proxy setup from our error messages. I think errors relating to the network fetch should use log.Errorf (without log.V(1) - that just makes it harder to see because no one knows which vmodule options to turn on).

Parsing could go either way - in theory the parse error would not contain any information that the client didn't just give to us, so it would be safe to return as an error detail, but it's probably better to play it safe and leave the details in the log. This would be a better use for log.V(1) so that a malicious client can't easily spam the logs.
</blockquote></details>

If we log it here, we would be double logging since we are also logging in `pgwire/auth_methods.go`
https://github.com/cockroachdb/cockroach/blob/b8ba30a6ee2a76ecf9f3dfc2f5c7665b84fb1ce4/pkg/sql/pgwire/auth_methods.go#L783-L786
Incase when we don't want to send back these details to client which are then logged with `c.LogAuthFailed()`, I am sending only the authLog back to client. c.LogAuthFailed() has the option of logging only unlogged errors but I don't have access to `AuthConn` object in `authentication_jwt.go` which can configure this parameter. The `AuthConn` object can be added but I thought of handling this in the higher layer of pgwire inline with previous comment: 
[https://reviewable.io/reviews/cockroachdb/cockroach/117054#-Nne7jnh5oeb9srWuZw2](https://reviewable.io/reviews/cockroachdb/cockroach/117054#-Nne7jnh5oeb9srWuZw2)

Also in pgwire, other authentication errors are not behind log.V(1), so I am assuming we were not focussed about malicious clients or overlogging. Only level 2 logs are being used for logging some info logs. LMK if we would still need this because our use case here is different from pwd or cert authentication.
___
*[`pkg/sql/pgwire/auth_methods.go` line 789 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/123697#-NxNVQs850pBOLCDo3-8:-NxShXER68-yCFBXBM3r:b-cm9b9o) ([raw file](https://github.com/cockroachdb/cockroach/blob/806b46e3d118c405261b78d2e459bb19356cc0ec/pkg/sql/pgwire/auth_methods.go#L789)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yes, it can be (although it depends on what the "unsafe" parts are). For example, usernames are generally marked as redacted and should not be sent to sentry, but they could be appropriate to return in an error message.
</blockquote></details>

`errors.WithDetailf()` mentions that error details  is supposed to contain PII data.
```
WithDetail decorates an error with a textual detail. The detail may contain PII and thus will not reportable. The suggested use case for detail is to augment errors with information useful for debugging.
```
Since we have the detailed error here which contains token related information which client itself sends, it makes sense to send back these to client, anyways sentry would be redacting these.
I think specifically for `c.LogAuthFailed()` we should add back the detailed error(like from http clients which might contain data sql client did not send us), so that it could be logged. I have made these changes here. 
___
*[`pkg/util/httputil/client.go` line 46 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/123697#-NxNVcazCCi3xE2K0tmE:-NxSl3uA5t99J-fXt_gl:b-rh0ewj) ([raw file](https://github.com/cockroachdb/cockroach/blob/806b46e3d118c405261b78d2e459bb19356cc0ec/pkg/util/httputil/client.go#L46)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think custom CAs are generally used with a new empty cert pool, not SystemCertPool.
</blockquote></details>

updated to use NewCertPool()
___
*[`pkg/ccl/jwtauthccl/authentication_jwt_test.go` line 797 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/123697#-NxNOpv2Bpu1O_1hE_Fk:-NxTXpjG3-aVSzVDv404:b-1d263c) ([raw file](https://github.com/cockroachdb/cockroach/blob/806b46e3d118c405261b78d2e459bb19356cc0ec/pkg/ccl/jwtauthccl/authentication_jwt_test.go#L797)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

try using `testutils.TestingHook` for this pattern just to keep things consistent.
</blockquote></details>

updated to use `testutils.TestingHook`
___
*[`pkg/sql/pgwire/auth.go` line 556 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/123697#-NxNV1hBEfFduTXn7rPX:-NxShKatD1tK2fwboOnK:b-r183vz) ([raw file](https://github.com/cockroachdb/cockroach/blob/806b46e3d118c405261b78d2e459bb19356cc0ec/pkg/sql/pgwire/auth.go#L556)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I don't think so - redaction is about what can be sent to CRL. For error messages we generally want to either send the whole thing to the client or none of it.
</blockquote></details>

noted. sending the logged error as is, the handling is done is `auth_methods.go` directly.

**Review by souravcrl — COMMENTED** [review]
I updated `auth_methods` logging, since I found we were redacting in `LogAuthFailed` called inside authenticator fn. If we put the error in details part  of errors.WithDetailf or had the error from http at a different level only the first level of error is logged. So I performed join operation on errors for logging. I have updated commit message to reflect that

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/123697#-:-NxWdGVt7yvXqtxMeYiw:b-rnl56n)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @BabuSrithar, @bdarnell, and @dhartunian)

___
*[`pkg/ccl/jwtauthccl/authentication_jwt.go` line 189 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/123697#gh-1593988975:-NxWd1ho5Qkayrtn0ZK4:b-b44ve8) ([raw file](https://github.com/cockroachdb/cockroach/blob/806b46e3d118c405261b78d2e459bb19356cc0ec/pkg/ccl/jwtauthccl/authentication_jwt.go#L189)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

If we log it here, we would be double logging since we are also logging in `pgwire/auth_methods.go`
https://github.com/cockroachdb/cockroach/blob/b8ba30a6ee2a76ecf9f3dfc2f5c7665b84fb1ce4/pkg/sql/pgwire/auth_methods.go#L783-L786
Incase when we don't want to send back these details to client which are then logged with `c.LogAuthFailed()`, I am sending only the authLog back to client. c.LogAuthFailed() has the option of logging only unlogged errors but I don't have access to `AuthConn` object in `authentication_jwt.go` which can configure this parameter. The `AuthConn` object can be added but I thought of handling this in the higher layer of pgwire inline with previous comment: 
[https://reviewable.io/reviews/cockroachdb/cockroach/117054#-Nne7jnh5oeb9srWuZw2](https://reviewable.io/reviews/cockroachdb/cockroach/117054#-Nne7jnh5oeb9srWuZw2)

Also in pgwire, other authentication errors are not behind log.V(1), so I am assuming we were not focussed about malicious clients or overlogging. Only level 2 logs are being used for logging some info logs. LMK if we would still need this because our use case here is different from pwd or cert authentication.
</blockquote></details>

We also have a cluster setting for logging auth errors: `server.auth_log.sql_sessions.enabled` which needs to be set to true

**Review by bdarnell — APPROVED** [review]
> Reviewed 4 of 10 files at r4, 3 of 3 files at r5, 6 of 6 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/123697)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @BabuSrithar, @dhartunian, and @souravcrl)

___
*[`pkg/ccl/jwtauthccl/authentication_jwt.go` line 142 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/123697#-NxY9WzmD4MUc_IA0REY:-NxY9WzmD4MUc_IA0REZ:b-8n4eti) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb57f5cb94bd2b22cad1a21f53c64402ab170f6b/pkg/ccl/jwtauthccl/authentication_jwt.go#L142)):*
> ```Go
> 	tokenBytes []byte,
> 	identMap *identmap.Conf,
> ) (detailedErrorMsg []byte, authError error) {
> ```

Why is detailedErrorMsg a `[]byte` instead of a `string`? or even another `error`?
___
*[`pkg/ccl/jwtauthccl/authentication_jwt.go` line 333 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/123697#-NxYEbPS77SetQxy-pbY:-NxYEbPS77SetQxy-pbZ:b-fc2098) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb57f5cb94bd2b22cad1a21f53c64402ab170f6b/pkg/ccl/jwtauthccl/authentication_jwt.go#L333)):*
> ```Go
> var getHttpResponse = func(ctx context.Context, url string, authenticator *jwtAuthenticator) ([]byte, error) {
> 	defaultTimeout := httputil.StandardHTTPTimeout
> 	httpClient := httputil.NewClientWithTimeoutsCustomCA(
> ```

This reparses the custom CA every time, which seems kind of expensive. I don't think we have to fix it in this PR (do we have a customer planning to use custom CAs?) but we should make a note to cache the httpClient somewhere (could do it in a callback attached to the cluster setting). 
___
*[`pkg/ccl/jwtauthccl/settings.go` line 91 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/123697#-NxYEHRf4QhxsmfU8jHL:-NxYEHRf4QhxsmfU8jHM:bjfaizz) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb57f5cb94bd2b22cad1a21f53c64402ab170f6b/pkg/ccl/jwtauthccl/settings.go#L91)):*
> ```Go
> 	settings.ApplicationLevel,
> 	JWTAuthClientCustomCASettingName,
> 	"sets the custom root CA (appended to system's default CAs) for verifying "+
> ```

"Appended to system's default CAs" is not accurate now that you're not using SystemCertPool below. I'd just remove the parenthetical.

**Review by souravcrl — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/123697)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @BabuSrithar, @bdarnell, and @dhartunian)

___
*[`pkg/ccl/jwtauthccl/authentication_jwt.go` line 142 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/123697#-NxY9WzmD4MUc_IA0REY:-NxYPW573bqNwiS8oxDA:bb03wds) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb57f5cb94bd2b22cad1a21f53c64402ab170f6b/pkg/ccl/jwtauthccl/authentication_jwt.go#L142)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why is detailedErrorMsg a `[]byte` instead of a `string`? or even another `error`?
</blockquote></details>

we cannot send as `error` because lint doesn't allow it inside a block where err != nil, but one of the returned variables is nil.
Reconsidered this and updated to use string to avoid unnecessary typecasts.
___
*[`pkg/ccl/jwtauthccl/authentication_jwt.go` line 333 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/123697#-NxYEbPS77SetQxy-pbY:-NxYdif-BS1VIZSBmfiy:be7ugew) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb57f5cb94bd2b22cad1a21f53c64402ab170f6b/pkg/ccl/jwtauthccl/authentication_jwt.go#L333)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This reparses the custom CA every time, which seems kind of expensive. I don't think we have to fix it in this PR (do we have a customer planning to use custom CAs?) but we should make a note to cache the httpClient somewhere (could do it in a callback attached to the cluster setting).
</blockquote></details>

the proxy server client is using may have custom CA requirement. adding a note to cache the http client in authenticator which could be set on change for customCA cluster setting.
___
*[`pkg/ccl/jwtauthccl/settings.go` line 91 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/123697#-NxYEHRf4QhxsmfU8jHL:-NxYgF0e7BjZRUykGwmY:bo2slgm) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb57f5cb94bd2b22cad1a21f53c64402ab170f6b/pkg/ccl/jwtauthccl/settings.go#L91)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

"Appended to system's default CAs" is not accurate now that you're not using SystemCertPool below. I'd just remove the parenthetical.
</blockquote></details>

remove the part in parentheses

**Comment by BabuSrithar** [comment]
*[`pkg/util/httputil/client.go` line 74 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/123697#-Nxkfckd-AZToPh_Io3z:-Nxkfckd-AZToPh_Io4-:b-2cclos) ([raw file](https://github.com/cockroachdb/cockroach/blob/d526a7af3c071d72ebab6f6d8c2b7ffc86a13ccf/pkg/util/httputil/client.go#L74)):*
> ```Go
> 
> 			// Add our custom CA.
> 			TLSClientConfig: tlsConf,
> ```

@bdarnell @souravcrl  - I am wondering if custom-ca is a relevant thing in case of JWT auth. The only place I see this being useful is JWT server having slef-signed cert. Even in that case, can't the self signed cert be imported in to system's cert store ?

**Review by bdarnell — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/123697)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @BabuSrithar, @dhartunian, and @souravcrl)

___
*[`pkg/util/httputil/client.go` line 74 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/123697#-Nxkfckd-AZToPh_Io3z:-NxmCB1i78dJS3dVUWVA:bv4xlgq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e053c5a0391849586b64f5ba0b0424f31433a0a8/pkg/util/httputil/client.go#L74)):*
<details><summary><i>Previously, BabuSrithar (BabuSrithar) wrote…</i></summary><blockquote>

@bdarnell @souravcrl  - I am wondering if custom-ca is a relevant thing in case of JWT auth. The only place I see this being useful is JWT server having slef-signed cert. Even in that case, can't the self signed cert be imported in to system's cert store ?
</blockquote></details>

Yes, this is useful when the JWT server has a cert that is either self-signed or comes from a private CA. You could add the private CA to the system cert store in this case (and I think most of the time it would be a good thing), but you might not want to. The system cert store is generally intended to store CAs for the public web. Mixing private CAs here might be undesirable from either direction. It would give your private CA the ability to impersonate public websites (probably not a problem in the context of a database server), and more worryingly it could give public CAs the ability to impersonate your private websites (this matters to the kinds of organizations that have to worry about high-level nation-state security threats). 

I think the question is whether we have customer feedback asking for this. I assumed we did if Sourav went to the trouble of adding it, but if not I think we could remove it and see if the system cert store alone can be a good solution.

**Review by souravcrl — COMMENTED** [review]
updated to remove the cluster setting for custom CA. PTAL @BabuSrithar 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/123697#-:-NxnGfKC2nbUgsxcTcnt:b-dz8nos)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @BabuSrithar, @bdarnell, and @dhartunian)

___
*[`pkg/ccl/jwtauthccl/authentication_jwt.go` line 333 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/123697#-NxnGZyR2MjPTVaU4xc7:-NxnGZyR2MjPTVaU4xc8:bs296uv) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb1923b8f490192b5d346a4024a90888cbb52551/pkg/ccl/jwtauthccl/authentication_jwt.go#L333)):*
> ```Go
> 	// and other http client cluster settings as re parsing the custom CA every
> 	// time is expensive
> 	httpClient := httputil.NewClientWithTimeout(httputil.StandardHTTPTimeout)
> ```

sticking to standard http timeout of 3 seconds here.
___
*[`pkg/util/httputil/client.go` line 74 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/123697#-Nxkfckd-AZToPh_Io3z:-NxnEWo_Ai4fh7r6hy5k:b-xyeli8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e053c5a0391849586b64f5ba0b0424f31433a0a8/pkg/util/httputil/client.go#L74)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yes, this is useful when the JWT server has a cert that is either self-signed or comes from a private CA. You could add the private CA to the system cert store in this case (and I think most of the time it would be a good thing), but you might not want to. The system cert store is generally intended to store CAs for the public web. Mixing private CAs here might be undesirable from either direction. It would give your private CA the ability to impersonate public websites (probably not a problem in the context of a database server), and more worryingly it could give public CAs the ability to impersonate your private websites (this matters to the kinds of organizations that have to worry about high-level nation-state security threats). 

I think the question is whether we have customer feedback asking for this. I assumed we did if Sourav went to the trouble of adding it, but if not I think we could remove it and see if the system cert store alone can be a good solution.
</blockquote></details>

removed the additional cluster setting and code for custom/private CA and sticking to system cert store solution. I have created a ticket with link to this discussion for future reference: https://cockroachlabs.atlassian.net/browse/CRDB-38682

**Comment by souravcrl** [comment]
TFTR!

bors r+

**Review by BabuSrithar — COMMENTED** [review]
> Reviewed 1 of 2 files at r3, 4 of 7 files at r9, 3 of 3 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/123697)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @dhartunian)

**Pull request merged**